README.rst
=======
pywinio
=======

.. image:: https://img.shields.io/pypi/v/pywinio.svg
    :target: https://pypi.python.org/pypi/pywinio

.. image:: https://ci.appveyor.com/api/projects/status/github/starofrainnight/pywinio?svg=true
    :target: https://ci.appveyor.com/project/starofrainnight/pywinio

A wrapper library for WinIO

Thanks for the great WinIO library which comes from Yariv Kaplan.

The WinIO library binary distribution will be accessed from winiobinary package.

NOTE: This library has been renamed from 'rabird.winio' to 'pywinio' after 0.3.0 .

* License: Apache-2.0

Usage
-----

* Keyboard Emulation

 NOTICE: You must have a ps/2 keyboard connected to your PC if you test on win10 or later, otherwise this sample won't works!

::

    import pywinio
    import time
    import atexit

    # KeyBoard Commands
    # Command port
    KBC_KEY_CMD	= 0x64
    # Data port
    KBC_KEY_DATA = 0x60

    __winio = None

    def __get_winio():
    	global __winio

    	if __winio is None:
    		__winio = pywinio.WinIO()
    		def __clear_winio():
    			global __winio
    			__winio = None
    		atexit.register(__clear_winio)

    	return __winio

    def wait_for_buffer_empty():
    	'''
    	Wait keyboard buffer empty
    	'''

    	winio = __get_winio()

    	dwRegVal = 0x02
    	while (dwRegVal & 0x02):
    		dwRegVal = winio.get_port_byte(KBC_KEY_CMD)

    def key_down(scancode):
    	winio = __get_winio()

    	wait_for_buffer_empty();
    	winio.set_port_byte(KBC_KEY_CMD, 0xd2);
    	wait_for_buffer_empty();
    	winio.set_port_byte(KBC_KEY_DATA, scancode)

    def key_up(scancode):
    	winio = __get_winio()

    	wait_for_buffer_empty();
    	winio.set_port_byte( KBC_KEY_CMD, 0xd2);
    	wait_for_buffer_empty();
    	winio.set_port_byte( KBC_KEY_DATA, scancode | 0x80);

    def key_press(scancode, press_time = 0.2):
    	key_down( scancode )
    	time.sleep( press_time )
    	key_up( scancode )


    # Press 'A' key
    # Scancodes references : https://www.win.tue.nl/~aeb/linux/kbd/scancodes-1.html
    key_press(0x1E)


Privilege Requirements
-----------------------

WinIo requires administrative privileges to run properly. This can be achieved by:

* Using WinIo from a service running as LocalSystem (SE_LOAD_DRIVER_NAME privilege must be explicitly enabled).
* Embedding a manifest file in the application that requests privilege elevation.
* Requiring the user to choose the "Run as Administrator" option when launching the application.

The included C# samples demonstrate using an embedded manifest file to request privilege elevation.

Driver Signing Requirements on 64-bit Systems
---------------------------------------------

64-bit versions of Windows only load device drivers that are signed by a code signing certificate issued by a public CA such as Verisign, Thawte, etc. WinIo64.sys must not be deployed on production machines unless a code signing certificate is obtained and used to sign this file. The bundled copy of WinIo64.sys is signed with a self-signed certificate and can only be used on development/test machines with Windows running in a special "test" mode. In order to use the bundled version of WinIo64.sys, you need to take the following steps:

* Open an elevated command window by right-clicking the icon and clicking "Run as Administrator".
* Type the following command to enable test-signing:

 ::

  bcdedit.exe /set TESTSIGNING ON

* Reboot the machine

For more information on Windows driver signing requirements, please refer to http://www.microsoft.com/whdc/winlogo/drvsign/kmcs_walkthrough.mspx.

Licenses
---------------------------------------------

This library was licensed under Apache-2.0 which already descripted in LICENSE

WinIO library's license descripted in LICENSE-WINIO


CHANGES.rst
Changes
=======


0.2.6 (2018-07-20)
---------------------

* Claim this library renamed to pywinio!

0.2.5 (2018-07-20)
---------------------

* Removed support of python 2.x

0.2.4 (2018-06-01)
---------------------

* Supported automatic deploy by appveyor

0.2.0 (2018-01-21)
---------------------

* Depends on winiobinary project, so we don't need to download WinIo binaries during setup.py

0.1.9 (2015-12-02)
---------------------

* Can't install from packaged zip file due to missing rabird_bootstrap.py

0.1.8 (2015-09-09)
---------------------

* Fixed not working with 64bits os

0.1.0 (2015-08-18)
---------------------

* Fixed not working with python3

0.0.9 (2015-08-18)
---------------------

* Initial release


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for pywinio."""


tests/test_pywinio.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `pywinio` package."""

import pytest

from click.testing import CliRunner

from pywinio import pywinio  # noqa


def test_fake():
    pass


tests/test_rabird_winio.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `rabird.winio` package."""

import rabird.winio


def test_rabird_winio():
    """
    Test for rabird.winio
    """

    assert True


ci/executor.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import sys
import click
import subprocess


def run(cmd):
    click.echo("RUN: %s" % cmd)
    return subprocess.check_call(cmd, shell=True)


def run_path(cmd):
    return run("%s %s" % (sys.executable, cmd))


def run_module(cmd):
    return run("%s -m %s" % (sys.executable, cmd))


@click.group()
def main():
    """Execute CI actions"""

    pass


@main.command()
@click.option('-e', '--env', multiple=True, default=['default'],
              type=click.Choice(['default', 'mypy', 'flake8']))
def test(env):
    """Test the project
    """
    envs = env
    for env in envs:
        if env == 'default':
            run_module("pip install -r./requirements_dev.txt")
            run_module("pip install -r./requirements.txt")
            run_path("setup.py test")
        elif env == 'mypy':
            run_module("pip install mypy")
            run_module("mypy -m pywinio --ignore-missing-imports")
        elif env == 'flake8':
            run_module("pip install flake8")
            run_module("flake8 pywinio tests")


@main.command()
def pack():
    """Build distructbution
    """

    run_path("setup.py sdist --formats=zip")


@main.command()
@click.pass_context
def deploy(ctx):
    """Deploy this project"""

    ctx.invoke(pack)
    run_module("pip install twine")
    run_module("twine upload dist/*")


if __name__ == '__main__':
    main()


pywinio/pywinio.py
import os
import ctypes
import win32con
import win32file
import winioctlcon
import pywintypes
import win32service
import struct
import winiobinary

# Define the various device type values.  Note that values used by Microsoft
# Corporation are in the range 0-32767, and 32768-65535 are reserved for use
# by customers.

FILE_DEVICE_WINIO = 0x00008010

# Macro definition for defining IOCTL and FSCTL function control codes.
# Note that function codes 0-2047 are reserved for Microsoft Corporation,
# and 2048-4095 are reserved for customers.

WINIO_IOCTL_INDEX = 0x810

# Define our own private IOCTL

IOCTL_WINIO_MAPPHYSTOLIN = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                                WINIO_IOCTL_INDEX,
                                                winioctlcon.METHOD_BUFFERED,
                                                winioctlcon.FILE_ANY_ACCESS)

IOCTL_WINIO_UNMAPPHYSADDR = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                                 WINIO_IOCTL_INDEX + 1,
                                                 winioctlcon.METHOD_BUFFERED,
                                                 winioctlcon.FILE_ANY_ACCESS)

IOCTL_WINIO_ENABLEDIRECTIO = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                                  WINIO_IOCTL_INDEX + 2,
                                                  winioctlcon.METHOD_BUFFERED,
                                                  winioctlcon.FILE_ANY_ACCESS)

IOCTL_WINIO_DISABLEDIRECTIO = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                                   WINIO_IOCTL_INDEX + 3,
                                                   winioctlcon.METHOD_BUFFERED,
                                                   winioctlcon.FILE_ANY_ACCESS)

IOCTL_WINIO_READPORT = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                            WINIO_IOCTL_INDEX + 4,
                                            winioctlcon.METHOD_BUFFERED,
                                            winioctlcon.FILE_ANY_ACCESS)

IOCTL_WINIO_WRITEPORT = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                             WINIO_IOCTL_INDEX + 5,
                                             winioctlcon.METHOD_BUFFERED,
                                             winioctlcon.FILE_ANY_ACCESS)


# Error Codes
EC_FILE_NOT_FOUND = 2
EC_SERVICE_FILE_NOT_FOUND = 3
EC_SERVICE_NOT_INSTALLED = 1060
EC_SERVICE_NOT_STARTED = 1062
EC_SERVICE_EXISTED = 1073


def NotInitializedError(RuntimeError):
    pass


def InvalidArgumentError(ValueError):
    pass


def DeviceIoControlError(RuntimeError):
    pass


class tagPhysStruct(ctypes.Structure):
    _fields_ = [
        ('dwPhysMemSizeInBytes', ctypes.c_ulonglong),
        ('pvPhysAddress', ctypes.c_ulonglong),
        ('PhysicalMemoryHandle', ctypes.c_ulonglong),
        ('pvPhysMemLin', ctypes.c_ulonglong),
        ('pvPhysSection', ctypes.c_ulonglong),
    ]
    _pack_ = 1


class tagPortStruct(ctypes.Structure):
    _fields_ = [
        ('wPortAddr', ctypes.c_ushort),
        ('dwPortVal', ctypes.c_ulong),
        ('bSize', ctypes.c_ubyte),
    ]
    _pack_ = 1


def _ensure_initialized(method):
    def wrapper(self, *args, **kwargs):
        if not self.dll_is_initialized:
            raise NotInitializedError()
        return method(self, *args, **kwargs)
    return wrapper


class WinIO(object):
    # Rewrote __new__ for Singleton design.

    def __new__(cls, *args, **kw):
        if not hasattr(cls, '_instance'):
            orig = super(WinIO, cls)
            cls._instance = orig.__new__(cls, *args, **kw)
        return cls._instance

    def __init__(self):
        self.dll_is_initialized = False
        self.hSCManager = win32service.OpenSCManager(
            None, None, win32service.SC_MANAGER_ALL_ACCESS)
        self.hDriver = None
        self.__is_need_uninstall_driver = True
        self.__initialize()

    def __stop_driver(self):
        hService = None

        hService = win32service.OpenService(
            self.hSCManager, 'WINIO', win32service.SERVICE_ALL_ACCESS)
        try:
            win32service.ControlService(
                hService, win32service.SERVICE_CONTROL_STOP)
        except pywintypes.error as e:
            if e.winerror == EC_SERVICE_NOT_STARTED:
                return
            else:
                raise e

    @_ensure_initialized
    def __get_port_value(self, wPortAddr, bSize):
        if bSize not in [1, 2, 4]:
            raise InvalidArgumentError(
                'Argument "bSize" only accept 1, 2, 4. Current value : %s!' % (
                    str(bSize)))

        # If this is a 64 bit OS, we must use the driver to access I/O ports
        # even if the application is 32 bit
        if self.__is_64bit_os():
            PortStruct = tagPortStruct()
            PortStruct.bSize = bSize
            PortStruct.wPortAddr = wPortAddr

            return struct.unpack('@L',
                                 win32file.DeviceIoControl(
                                     self.hDriver,
                                     IOCTL_WINIO_READPORT,
                                     memoryview(PortStruct),
                                     4))[0]
        else:
            if bSize == 1:
                return ctypes.c_ubyte(ctypes.cdll.msvcrt._inp(
                    ctypes.c_ushort(wPortAddr))).value
            elif bSize == 2:
                return ctypes.c_ushort(ctypes.cdll.msvcrt._inpw(
                    ctypes.c_ushort(wPortAddr))).value
            elif bSize == 4:
                return ctypes.c_ulong(ctypes.cdll.msvcrt._inpd(
                    ctypes.c_ushort(wPortAddr))).value

    def get_port_byte(self, wPortAddr):
        return self.__get_port_value(wPortAddr, 1)

    def get_port_word(self, wPortAddr):
        return self.__get_port_value(wPortAddr, 2)

    def get_port_dword(self, wPortAddr):
        return self.__get_port_value(wPortAddr, 4)

    @_ensure_initialized
    def __set_port_value(self, wPortAddr, dwPortVal, bSize):
        # If this is a 64 bit OS, we must use the driver to access I/O ports
        # even if the application is 32 bit
        if self.__is_64bit_os():
            PortStruct = tagPortStruct()
            PortStruct.bSize = bSize
            PortStruct.dwPortVal = dwPortVal
            PortStruct.wPortAddr = wPortAddr

            win32file.DeviceIoControl(
                self.hDriver,
                IOCTL_WINIO_WRITEPORT,
                memoryview(PortStruct),
                4)
        else:
            if bSize == 1:
                ctypes.cdll.msvcrt._outp(ctypes.c_ushort(
                    wPortAddr), ctypes.c_int(dwPortVal))
            elif bSize == 2:
                ctypes.cdll.msvcrt._outpw(ctypes.c_ushort(
                    wPortAddr), ctypes.c_ushort(dwPortVal))
            elif bSize == 4:
                ctypes.cdll.msvcrt._outpd(ctypes.c_ushort(
                    wPortAddr), ctypes.c_ulong(dwPortVal))

    def set_port_byte(self, wPortAddr, dwPortVal):
        return self.__set_port_value(wPortAddr, dwPortVal, 1)

    def set_port_word(self, wPortAddr, dwPortVal):
        return self.__set_port_value(wPortAddr, dwPortVal, 2)

    def set_port_dword(self, wPortAddr, dwPortVal):
        return self.__set_port_value(wPortAddr, dwPortVal, 4)

    @_ensure_initialized
    def map_phys_to_lin(self, PhysStruct):
        if (not win32file.DeviceIoControl(
                self.hDriver,
                IOCTL_WINIO_MAPPHYSTOLIN,
                memoryview(PhysStruct),
                memoryview(PhysStruct))):
            raise DeviceIoControlError("Failed on IOCTL_WINIO_MAPPHYSTOLIN")

        return PhysStruct.pvPhysMemLin

    @_ensure_initialized
    def unmap_physical_memory(self, PhysStruct):
        if (not win32file.DeviceIoControl(
                self.hDriver,
                IOCTL_WINIO_UNMAPPHYSADDR,
                memoryview(PhysStruct),
                None)):
            raise DeviceIoControlError("Failed on IOCTL_WINIO_UNMAPPHYSADDR")

    @_ensure_initialized
    def get_phys_long(self, pbPhysAddr):
        PhysStruct = tagPhysStruct()
        PhysStruct.pvPhysAddress = pbPhysAddr
        PhysStruct.dwPhysMemSizeInBytes = 4

        pdwLinAddr = self.map_phys_to_lin(PhysStruct)
        dwLinAddrRaw = bytes(
            (ctypes.c_char * 4).from_address(pdwLinAddr))
        dwPhysVal = struct.unpack('@L', dwLinAddrRaw)

        self.unmap_physical_memory(PhysStruct)

        return dwPhysVal

    @_ensure_initialized
    def set_phys_long(self, pbPhysAddr, dwPhysVal):
        PhysStruct = tagPhysStruct()
        PhysStruct.pvPhysAddress = pbPhysAddr
        PhysStruct.dwPhysMemSizeInBytes = 4

        pdwLinAddr = self.map_phys_to_lin(PhysStruct)
        if (pdwLinAddr == 0):
            raise ValueError("Linear address invalid!")

        data = ctypes.cast(
            ctypes.c_void_p(pdwLinAddr),
            ctypes.POINTER(ctypes.c_ulong))
        data.contents.value = dwPhysVal

        self.unmap_physical_memory(PhysStruct)

    def uninstall_driver(self):
        hService = None
        pServiceConfig = None
        # dwBytesNeeded = 0
        # cbBufSize = 0

        try:
            self.__stop_driver()

            hService = win32service.OpenService(
                self.hSCManager, 'WINIO', win32service.SERVICE_ALL_ACCESS)

            # If QueryServiceConfig() can not return a correct config, it will
            # throw exception!
            pServiceConfig = win32service.QueryServiceConfig(hService)

            # If service is set to load automatically, don't delete it!
            # dwStartType
            if (pServiceConfig[1] == win32service.SERVICE_DEMAND_START):
                win32service.DeleteService(hService)

        except pywintypes.error as e:
            if e.winerror == EC_SERVICE_NOT_INSTALLED:
                return

            raise

    def install_driver(self, pszWinIoDriverPath, IsDemandLoaded):
        # Remove any previous instance of the driver
        self.uninstall_driver()

        # Install the driver
        if (IsDemandLoaded is True):
            demand_flags = win32service.SERVICE_DEMAND_START
        else:
            demand_flags = win32service.SERVICE_SYSTEM_START

        try:
            win32service.CreateService(
                self.hSCManager,
                "WINIO",
                "WINIO",
                win32service.SERVICE_ALL_ACCESS,
                win32service.SERVICE_KERNEL_DRIVER,
                demand_flags,
                win32service.SERVICE_ERROR_NORMAL,
                pszWinIoDriverPath,
                None,
                0,
                None,
                None,
                None)
        except pywintypes.error as e:
            if e.winerror == EC_SERVICE_EXISTED:  # Service existed!
                return True

            raise

    def __start_driver(self):
        hService = None
        hService = win32service.OpenService(
            self.hSCManager, "WINIO", win32service.SERVICE_ALL_ACCESS)
        win32service.StartService(hService, None)

    def __is_64bit_os(self):
        if 'PROCESSOR_ARCHITEW6432' in os.environ:
            if os.environ['PROCESSOR_ARCHITEW6432'] == 'AMD64':
                return True

        if "64" in os.environ["PROCESSOR_ARCHITECTURE"]:
            return True

        return False

    def __get_driver_file_path(self):
        # Find the data directory
        winio_binary_data_dir = winiobinary.get_data_dir()

        if self.__is_64bit_os():
            result = os.path.join(winio_binary_data_dir, 'WinIo64.sys')
        else:
            result = os.path.join(winio_binary_data_dir, 'WinIo32.sys')

        result = os.path.normpath(os.path.realpath(result))

        return result

    def __initialize(self):
        # Try to open the winio device.
        hDriver = None
        try:
            hDriver = win32file.CreateFile(
                r"\\.\WINIO",
                win32con.GENERIC_READ | win32con.GENERIC_WRITE,
                0,
                None,
                win32con.OPEN_EXISTING,
                win32con.FILE_ATTRIBUTE_NORMAL,
                None)

            # Driver already installed in system as service by user, we must
            # not uninstall it.
            self.__is_need_uninstall_driver = False
        except pywintypes.error as e:
            if e.winerror != EC_FILE_NOT_FOUND:
                raise e

        # If the driver is not running, install it
        if hDriver is None:
            self.install_driver(self.__get_driver_file_path(), True)
            self.__start_driver()

            hDriver = win32file.CreateFile(
                r"\\.\WINIO",
                win32con.GENERIC_READ | win32con.GENERIC_WRITE,
                win32con.FILE_SHARE_READ | win32con.FILE_SHARE_WRITE,
                None,
                win32con.OPEN_EXISTING,
                win32con.FILE_ATTRIBUTE_NORMAL,
                None)

        # Enable I/O port access for this process if running on a 32 bit OS
        if not self.__is_64bit_os():
            win32file.DeviceIoControl(
                hDriver, IOCTL_WINIO_ENABLEDIRECTIO, None, 4)

        self.hDriver = hDriver
        self.dll_is_initialized = True

    def __finalize(self):
        if self.hDriver is not None:
            # Disable I/O port access if running on a 32 bit OS
            if not self.__is_64bit_os():
                win32file.DeviceIoControl(
                    self.hDriver, IOCTL_WINIO_DISABLEDIRECTIO, None, 4)

            self.hDriver = None

        if self.__is_need_uninstall_driver:
            self.uninstall_driver()

        self.dll_is_initialized = False

    def __del__(self):
        if self.dll_is_initialized:
            self.__finalize()


src/rabird/winio.py
import os
import sys
import ctypes
import atexit
import win32api
import win32con
import win32file
import winioctlcon
import pywintypes
import win32service
import locale
import logging
import ctypes
import struct
import winiobinary

__author__ = """Hong-She Liang"""
__email__ = 'starofrainnight@gmail.com'
__version__ = '0.2.5'

# Define the various device type values.  Note that values used by Microsoft
# Corporation are in the range 0-32767, and 32768-65535 are reserved for use
# by customers.

FILE_DEVICE_WINIO = 0x00008010

# Macro definition for defining IOCTL and FSCTL function control codes.
# Note that function codes 0-2047 are reserved for Microsoft Corporation,
# and 2048-4095 are reserved for customers.

WINIO_IOCTL_INDEX = 0x810

# Define our own private IOCTL

IOCTL_WINIO_MAPPHYSTOLIN = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                                WINIO_IOCTL_INDEX,
                                                winioctlcon.METHOD_BUFFERED,
                                                winioctlcon.FILE_ANY_ACCESS)

IOCTL_WINIO_UNMAPPHYSADDR = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                                 WINIO_IOCTL_INDEX + 1,
                                                 winioctlcon.METHOD_BUFFERED,
                                                 winioctlcon.FILE_ANY_ACCESS)

IOCTL_WINIO_ENABLEDIRECTIO = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                                  WINIO_IOCTL_INDEX + 2,
                                                  winioctlcon.METHOD_BUFFERED,
                                                  winioctlcon.FILE_ANY_ACCESS)

IOCTL_WINIO_DISABLEDIRECTIO = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                                   WINIO_IOCTL_INDEX + 3,
                                                   winioctlcon.METHOD_BUFFERED,
                                                   winioctlcon.FILE_ANY_ACCESS)

IOCTL_WINIO_READPORT = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                            WINIO_IOCTL_INDEX + 4,
                                            winioctlcon.METHOD_BUFFERED,
                                            winioctlcon.FILE_ANY_ACCESS)

IOCTL_WINIO_WRITEPORT = winioctlcon.CTL_CODE(FILE_DEVICE_WINIO,
                                             WINIO_IOCTL_INDEX + 5,
                                             winioctlcon.METHOD_BUFFERED,
                                             winioctlcon.FILE_ANY_ACCESS)


# Error Codes
EC_FILE_NOT_FOUND = 2
EC_SERVICE_FILE_NOT_FOUND = 3
EC_SERVICE_NOT_INSTALLED = 1060
EC_SERVICE_NOT_STARTED = 1062
EC_SERVICE_EXISTED = 1073


def NotInitializedError(RuntimeError):
    pass


def InvalidArgumentError(ValueError):
    pass


def DeviceIoControlError(RuntimeError):
    pass


class tagPhysStruct(ctypes.Structure):
    _fields_ = [
        ('dwPhysMemSizeInBytes', ctypes.c_ulonglong),
        ('pvPhysAddress', ctypes.c_ulonglong),
        ('PhysicalMemoryHandle', ctypes.c_ulonglong),
        ('pvPhysMemLin', ctypes.c_ulonglong),
        ('pvPhysSection', ctypes.c_ulonglong),
    ]
    _pack_ = 1


class tagPortStruct(ctypes.Structure):
    _fields_ = [
        ('wPortAddr', ctypes.c_ushort),
        ('dwPortVal', ctypes.c_ulong),
        ('bSize', ctypes.c_ubyte),
    ]
    _pack_ = 1


def _ensure_initialized(method):
    def wrapper(self, *args, **kwargs):
        if not self.dll_is_initialized:
            raise NotInitializedError()
        return method(self, *args, **kwargs)
    return wrapper


class WinIO(object):
    # Rewrote __new__ for Singleton design.

    def __new__(cls, *args, **kw):
        if not hasattr(cls, '_instance'):
            orig = super(WinIO, cls)
            cls._instance = orig.__new__(cls, *args, **kw)
        return cls._instance

    def __init__(self):
        self.dll_is_initialized = False
        self.hSCManager = win32service.OpenSCManager(
            None, None, win32service.SC_MANAGER_ALL_ACCESS)
        self.hDriver = None
        self.__is_need_uninstall_driver = True
        self.__initialize()

    def __stop_driver(self):
        hService = None
        ServiceStatus = None

        hService = win32service.OpenService(
            self.hSCManager, 'WINIO', win32service.SERVICE_ALL_ACCESS)
        try:
            ServiceStatus = win32service.ControlService(
                hService, win32service.SERVICE_CONTROL_STOP)
        except pywintypes.error as e:
            if e.winerror == EC_SERVICE_NOT_STARTED:
                return
            else:
                raise e

    @_ensure_initialized
    def __get_port_value(self, wPortAddr, bSize):
        if bSize not in [1, 2, 4]:
            raise InvalidArgumentError(
                'Argument "bSize" only accept 1, 2, 4. Current value : %s!' % (str(bSize)))

        # If this is a 64 bit OS, we must use the driver to access I/O ports even
        # if the application is 32 bit
        if self.__is_64bit_os():
            PortStruct = tagPortStruct()
            PortStruct.bSize = bSize
            PortStruct.wPortAddr = wPortAddr

            return struct.unpack('@L',
                                 win32file.DeviceIoControl(
                                     self.hDriver,
                                     IOCTL_WINIO_READPORT,
                                     memoryview(PortStruct),
                                     4))[0]
        else:
            if bSize == 1:
                return ctypes.c_ubyte(ctypes.cdll.msvcrt._inp(ctypes.c_ushort(wPortAddr))).value
            elif bSize == 2:
                return ctypes.c_ushort(ctypes.cdll.msvcrt._inpw(ctypes.c_ushort(wPortAddr))).value
            elif bSize == 4:
                return ctypes.c_ulong(ctypes.cdll.msvcrt._inpd(ctypes.c_ushort(wPortAddr))).value

    def get_port_byte(self, wPortAddr):
        return self.__get_port_value(wPortAddr, 1)

    def get_port_word(self, wPortAddr):
        return self.__get_port_value(wPortAddr, 2)

    def get_port_dword(self, wPortAddr):
        return self.__get_port_value(wPortAddr, 4)

    @_ensure_initialized
    def __set_port_value(self, wPortAddr, dwPortVal, bSize):
        # If this is a 64 bit OS, we must use the driver to access I/O ports even
        # if the application is 32 bit
        if self.__is_64bit_os():
            PortStruct = tagPortStruct()
            PortStruct.bSize = bSize
            PortStruct.dwPortVal = dwPortVal
            PortStruct.wPortAddr = wPortAddr

            win32file.DeviceIoControl(
                self.hDriver,
                IOCTL_WINIO_WRITEPORT,
                memoryview(PortStruct),
                4)
        else:
            if bSize == 1:
                ctypes.cdll.msvcrt._outp(ctypes.c_ushort(
                    wPortAddr), ctypes.c_int(dwPortVal))
            elif bSize == 2:
                ctypes.cdll.msvcrt._outpw(ctypes.c_ushort(
                    wPortAddr), ctypes.c_ushort(dwPortVal))
            elif bSize == 4:
                ctypes.cdll.msvcrt._outpd(ctypes.c_ushort(
                    wPortAddr), ctypes.c_ulong(dwPortVal))

    def set_port_byte(self, wPortAddr, dwPortVal):
        return self.__set_port_value(wPortAddr, dwPortVal, 1)

    def set_port_word(self, wPortAddr, dwPortVal):
        return self.__set_port_value(wPortAddr, dwPortVal, 2)

    def set_port_dword(self, wPortAddr, dwPortVal):
        return self.__set_port_value(wPortAddr, dwPortVal, 4)

    @_ensure_initialized
    def map_phys_to_lin(self, PhysStruct):
        if (not win32file.DeviceIoControl(
                self.hDriver,
                IOCTL_WINIO_MAPPHYSTOLIN,
                memoryview(PhysStruct),
                memoryview(PhysStruct))):
            raise DeviceIoControlError("Failed on IOCTL_WINIO_MAPPHYSTOLIN")

        return PhysStruct.pvPhysMemLin

    @_ensure_initialized
    def unmap_physical_memory(self, PhysStruct):
        if (not win32file.DeviceIoControl(
                self.hDriver,
                IOCTL_WINIO_UNMAPPHYSADDR,
                memoryview(PhysStruct),
                None)):
            raise DeviceIoControlError("Failed on IOCTL_WINIO_UNMAPPHYSADDR")

    @_ensure_initialized
    def get_phys_long(self, pbPhysAddr):
        PhysStruct = tagPhysStruct()
        PhysStruct.pvPhysAddress = pbPhysAddr
        PhysStruct.dwPhysMemSizeInBytes = 4

        pdwLinAddr = self.map_phys_to_lin(PhysStruct)
        dwLinAddrRaw = bytes(
            (ctypes.c_char * 4).from_address(pdwLinAddr))
        dwPhysVal = struct.unpack('@L', dwLinAddrRaw)

        self.unmap_physical_memory(PhysStruct)

        return dwPhysVal

    @_ensure_initialized
    def set_phys_long(self, pbPhysAddr, dwPhysVal):
        PhysStruct = tagPhysStruct()
        PhysStruct.pvPhysAddress = pbPhysAddr
        PhysStruct.dwPhysMemSizeInBytes = 4

        pdwLinAddr = self.map_phys_to_lin(PhysStruct)
        if (pdwLinAddr == 0):
            raise ValueError("Linear address invalid!")

        data = ctypes.cast(
            ctypes.c_void_p(pdwLinAddr),
            ctypes.POINTER(ctypes.c_ulong))
        data.contents.value = dwPhysVal

        self.unmap_physical_memory(PhysStruct)

    def uninstall_driver(self):
        hService = None
        pServiceConfig = None
        dwBytesNeeded = 0
        cbBufSize = 0

        try:
            self.__stop_driver()

            hService = win32service.OpenService(
                self.hSCManager, 'WINIO', win32service.SERVICE_ALL_ACCESS)

            # If QueryServiceConfig() can not return a correct config, it will
            # throw exception!
            pServiceConfig = win32service.QueryServiceConfig(hService)

            # If service is set to load automatically, don't delete it!
            # dwStartType
            if (pServiceConfig[1] == win32service.SERVICE_DEMAND_START):
                win32service.DeleteService(hService)

        except pywintypes.error as e:
            if e.winerror == EC_SERVICE_NOT_INSTALLED:
                return

            raise

    def install_driver(self, pszWinIoDriverPath, IsDemandLoaded):
        hService = None

        # Remove any previous instance of the driver
        self.uninstall_driver()

        # Install the driver
        if (IsDemandLoaded == True):
            demand_flags = win32service.SERVICE_DEMAND_START
        else:
            demand_flags = win32service.SERVICE_SYSTEM_START

        try:
            hService = win32service.CreateService(self.hSCManager,
                                                  "WINIO",
                                                  "WINIO",
                                                  win32service.SERVICE_ALL_ACCESS,
                                                  win32service.SERVICE_KERNEL_DRIVER,
                                                  demand_flags,
                                                  win32service.SERVICE_ERROR_NORMAL,
                                                  pszWinIoDriverPath,
                                                  None,
                                                  0,
                                                  None,
                                                  None,
                                                  None)
        except pywintypes.error as e:
            if e.winerror == EC_SERVICE_EXISTED:  # Service existed!
                return True

            raise

    def __start_driver(self):
        hService = None
        bResult = True

        hService = win32service.OpenService(
            self.hSCManager, "WINIO", win32service.SERVICE_ALL_ACCESS)
        win32service.StartService(hService, None)

    def __is_64bit_os(self):
        if 'PROCESSOR_ARCHITEW6432' in os.environ:
            if os.environ['PROCESSOR_ARCHITEW6432'] == 'AMD64':
                return True

        if "64" in os.environ["PROCESSOR_ARCHITECTURE"]:
            return True

        return False

    def __get_driver_file_path(self):
        # Find the data directory
        winio_binary_data_dir = winiobinary.get_data_dir()

        if self.__is_64bit_os():
            result = os.path.join(winio_binary_data_dir, 'WinIo64.sys')
        else:
            result = os.path.join(winio_binary_data_dir, 'WinIo32.sys')

        result = os.path.normpath(os.path.realpath(result))

        return result

    def __initialize(self):
        # Try to open the winio device.
        hDriver = None
        try:
            hDriver = win32file.CreateFile(r"\\.\WINIO",
                                           win32con.GENERIC_READ | win32con.GENERIC_WRITE,
                                           0,
                                           None,
                                           win32con.OPEN_EXISTING,
                                           win32con.FILE_ATTRIBUTE_NORMAL,
                                           None)

            # Driver already installed in system as service by user, we must
            # not uninstall it.
            self.__is_need_uninstall_driver = False
        except pywintypes.error as e:
            if e.winerror != EC_FILE_NOT_FOUND:
                raise e

        # If the driver is not running, install it
        if hDriver is None:
            self.install_driver(self.__get_driver_file_path(), True)
            self.__start_driver()

            hDriver = win32file.CreateFile(r"\\.\WINIO",
                                           win32con.GENERIC_READ | win32con.GENERIC_WRITE,
                                           win32con.FILE_SHARE_READ | win32con.FILE_SHARE_WRITE,
                                           None,
                                           win32con.OPEN_EXISTING,
                                           win32con.FILE_ATTRIBUTE_NORMAL,
                                           None)

        # Enable I/O port access for this process if running on a 32 bit OS
        if not self.__is_64bit_os():
            win32file.DeviceIoControl(
                hDriver, IOCTL_WINIO_ENABLEDIRECTIO, None, 4)

        self.hDriver = hDriver
        self.dll_is_initialized = True

    def __finalize(self):
        if self.hDriver is not None:
            # Disable I/O port access if running on a 32 bit OS
            if not self.__is_64bit_os():
                win32file.DeviceIoControl(
                    self.hDriver, IOCTL_WINIO_DISABLEDIRECTIO, None, 4)

            self.hDriver = None

        if self.__is_need_uninstall_driver:
            self.uninstall_driver()

        dll_is_initialized = False

    def __del__(self):
        if self.dll_is_initialized:
            self.__finalize()
